cpp/include/Ice/Buffer.h

```diff
@@ -82,6 +82,8 @@ namespace IceInternal
 
             bool empty() const { return !_size; }
 
+            bool ownsMemory() const noexcept { return _owned; }
```

Comment:
Should we only define this method if NDEBUG is not defined?

Reply:
I think that's fine as is, ownsMemory is very cheap. 

We could also throw an exception based on ownsMemory instead of asserting like I did in this PR.